src/main/java/pl/edu/mimuw/BigInt.java

```diff
@@ -20,38 +35,188 @@ public BigInt(int[] digits, boolean isPositive) {
+        r--;
+      frag = new int[r-l+1];
+      for (int i=l; i<=r; i++)
+        frag[i-l] = num[i];
+    }
+    return frag;
+  }
+
+  private int[] pow10(int[] num, int k) {
+    if (num.length == 1 && num[0] == 0)
+      return num;
+    int[] res = new int[num.length+k];
+    for (int i=0; i<k; i++)
+      res[i] = 0;
+    for (int i=0; i<num.length; i++)
+      res[i+k] = num[i];
+    return res;
+  }
+
+  private int[] karatsuba(int[] num1, int[] num2) {
+   int x;
+   int[] res, z0, z1, z2, left1, left2, right1, right2;
+   if (num1.length == 1 && num2.length == 1) {
+     x = num1[0] * num2[0];
+     if (x > 9) {
+       res = new int[2];
+       res[0] = x % 10;
+       res[1] = x / 10;
+     } else {
+       res = new int[1];
+       res[0] = x;
+     }
+   } else {
+     x = max(num1.length, num2.length) - 1;
+     left1 = cut(num1, 0, x/2);
+     right1 = cut(num1, x/2+1, x);
+     left2 = cut(num2, 0, x/2);
+     right2 = cut(num2, x/2+1, x);
+     z0 = karatsuba(left1, left2);
+     z2 = karatsuba(right1, right2);
+     z1 = auxSub(auxSub(karatsuba(auxAdd(left1, right1), auxAdd(left2, right2)), z0), z2);
+     res = auxAdd(auxAdd(pow10(z2, 2*(x/2+1)), pow10(z1, x/2+1)), z0);
+   }
+   return res;
+  }
+
   public BigInt add(BigInt other) {
-    throw new IllegalStateException("TODO task 2: implement addition");
+    BigInt res = new BigInt("0");
+    if (isPositive && other.isPositive) {
+      res.digits = auxAdd(digits, other.digits);
+      res.isPositive = true;
+    } else if (isPositive && !other.isPositive) {
+       if (auxCmp(digits, other.digits) < 0) {
+         res.digits = auxSub(other.digits, digits);
+         res.isPositive = false;
+       } else {
+         res.digits = auxSub(digits, other.digits);
+         res.isPositive = true;
+       }
+    } else if (!isPositive && other.isPositive) {
+       if (auxCmp(digits, other.digits) <= 0) {
+         res.digits = auxSub(other.digits, digits);
+         res.isPositive = true;
+       } else {
+         res.digits = auxSub(digits, other.digits);
+         res.isPositive = false;
+       }
+    } else {
+       res.digits = auxAdd(digits, other.digits);
+       res.isPositive = false;
+    }
+    return res;
   }
 
   public BigInt times(BigInt other) {
-    throw new IllegalStateException("TODO task 3: implement multiplication");
+    BigInt res = new BigInt("0");
+    if ((isPositive && other.isPositive)
+    || (!isPositive && !other.isPositive)
+    || (digits.length == 1 && digits[0] == 0)
+    || (other.digits.length == 1 && other.digits[0] == 0))
+      res.isPositive = true;
+    else
+      res.isPositive = false;
+    res.digits = karatsuba(digits, other.digits);
+    return res;
   }
 
   @Override
   public String toString() {
-    throw new IllegalStateException("TODO task 4: implement creating representation");
+    String number = Arrays.toString(digits).replace(",", "").replace("[", "").replace("]", "").replace(" ", "");
```

Comment:
zobacz na StringBuilder u mnie na stronie w ćw o Poly
to też jakiś sposób, ale nietpowy musze powiedziec